### PR TITLE
DO NOT LAND: for #9975: measure fullscreen performance

### PR DESCRIPTION
**Pull request by mcomella**
@MarcLeclair can you 1) review this code for correctness and 2) use this code to measure how long it takes to enter and exit fullscreen mode? The code won't be landing so just review for correctness, not other things like how easy is it to maintain. The steps to measure are:
- Install the build
- Go to a page, open and close fullscreen (to close, use the in-page button and back button)
- Execute `adb shell am start -n "org.mozilla.<app_id>/mozilla.telemetry.glean.debug.GleanDebugActivity" --ez logPings true --es sendPing metrics --es tagPings performance`, switching out the app ID, to send the measurements to the glean debug view
- Check the Glean Debug View for the result; feel free to send it to me if you want me to interpret
  - Note that due to the metric type we capture here (Timing Distribution), the values are bucketed rather than actual measurements

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by mcomella** [comment]
I tested on my GS5 on youtube.com:
- 319ms, 380ms, and 638ms to enter fullscreen, rounded to Glean bucketing
- 123ms, 615ms 2x to exit fullscreen, rounded to Glean bucketing

This doesn't match our responsiveness goals but as Eric mentioned https://github.com/mozilla-mobile/fenix/issues/9341#issuecomment-616951023:
> Existing investigation found a number of functional issues, however, the responsiveness of this portion of he UX is predominately controlled by the MSE JS players.

So I'm not sure how actionable this is.

**Comment by mcomella** [comment]
I'll wait for Marc to complete his analysis so we can compare the two devices and double-check with ecsmyth that there is nothing to do here.

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 1]
```diff
@@ -211,6 +213,8 @@ open class HomeActivity : LocaleAwareAppCompatActivity() {
     }
 
     final override fun onBackPressed() {
+        FullscreenTransitionMeasurement.onBackPressed() // DO NOT MOVE ANYTHING ABOVE HERE.
```
Comment (MarcLeclair) [inline app/src/main/java/org/mozilla/fenix/HomeActivity.kt:216]: is there a reason why we can't move anything above here? Just wondering.

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt [hunk 2]
```diff
@@ -219,6 +223,11 @@ open class HomeActivity : LocaleAwareAppCompatActivity() {
         super.onBackPressed()
     }
 
+    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
+        FullscreenTransitionMeasurement.onHomeActivityDispatchTouchEvent(ev) // DO NOT MOVE ANYTHING ABOVE HERE.
```
Comment (MarcLeclair) [inline app/src/main/java/org/mozilla/fenix/HomeActivity.kt:227]: same question as above, is there a reason why?

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 1 above)
Comment (mcomella) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:216]: If code is moved above here and it takes a long time, it won't be included in the measurement: we want the measurement to happen first.

I wonder if there's a more foolproof way to do this (e.g. with annotations?).

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 2 above)
Comment (mcomella) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:227]: Same answer. :)

File: app/src/main/java/org/mozilla/fenix/HomeActivity.kt (hunk 1 above)
Comment (MarcLeclair) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:216]: So, I think with anything like this, we shouldn't couple performance code with our app code. I was thinking more of an AOP way ( I'm thinking the way Spring does things) but looking around quickly I found this article: https://medium.com/dont-code-me-on-that/aspect-oriented-programming-in-kotlin-95ff8598913 which highlights the way we can do it in kotlin :) THoughts?

Comment (mcomella) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:216]: Oh interesting: that's pretty neat and avoids the footgun of someone moving code where they're not supposed to. Sure, let's aim for something like that in the future (though I'm going to do a little more reading on AOP to understand it more holistically).

Comment (MarcLeclair) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:216]: Yeah for sure, if we don't want to do this here, I think we can go ahead with this PR. It helps with less coupling. 😄

Comment (mcomella) [reply app/src/main/java/org/mozilla/fenix/HomeActivity.kt:216]: We're not intending to land this: sorry if that wasn't clear. The code is just here for you to take measurements.

**Comment by MarcLeclair** [comment]
So for the G5+ I'm getting  approximately 355ms to enter Fullscreen and ~1.4 seconds to exit full screen.

**Comment by mcomella** [comment]
That's all we need for this PR: I will complete the analysis and discussion in the comments following https://github.com/mozilla-mobile/fenix/issues/9341#issuecomment-616951023.

**Pull request closed without merging**